js/dydx.js

```diff
@@ -0,0 +1,664 @@
+            const marketKey = keys[i];
+            const market = markets[marketKey];
+            const id = this.safeString (market, 'market', marketKey);
+            const active = this.safeString (market, 'status') === 'ONLINE';
+            const baseId = this.safeString (market, 'baseAsset');
+            const quoteId = this.safeString (market, 'quoteAsset');
+            const base = this.safeCurrencyCode (baseId);
+            const quote = this.safeCurrencyCode (quoteId);
+            const settle = this.safeCurrencyCode (settleId);
+            const symbol = this.symbolDefine (base, quote, settle);
+            const stepSize = this.safeNumber (market, 'stepSize');
+            const tickSize = this.safeNumber (market, 'tickSize');
+            const minOrderSize = this.safeNumber (market, 'minOrderSize');
+            const type = this.safeString (market, 'type');
+            const spot = false;
+            const future = false;
+            const swap = (type === 'PERPETUAL');
+            const option = false;
+            const contract = future || swap;
+            const entry = {
+                'id': id,
+                'symbol': symbol,
+                'base': base,
+                'quote': quote,
+                'settle': settle,
+                'baseId': baseId,
+                'quoteId': quoteId,
+                'settleId': settleId,
+                'type': type,
+                'spot': spot,
+                'margin': false,
+                'future': future,
+                'swap': swap,
+                'option': option,
+                'active': active,
+                'contract': contract,
+                'linear': undefined,
+                'inverse': undefined,
+                'contractSize': undefined,
+                'expiry': undefined,
+                'expiryDatetime': undefined,
+                'strike': undefined,
+                'optionType': undefined,
+                'precision': {
+                    'price': tickSize,
+                    'amount': stepSize,
+                },
+                'limits': {
+                    'leverage': {
+                        'min': undefined,
+                        'max': undefined,
+                    },
+                    'amount': {
+                        'min': minOrderSize,
+                        'max': undefined,
+                    },
+                    'price': {
+                        'min': undefined,
+                        'max': undefined,
+                    },
+                    'cost': {
+                        'min': undefined,
+                        'max': undefined,
+                    },
+                },
+                'info': market,
+            };
+            result.push (entry);
+        }
+        return result;
+    }
+
+    symbolDefine (baseCode, quoteCode, settleCode = undefined, deliveryDate = undefined) { // TODO: can be unified method? users will use to avoid any mistakes
```

Comment:
I would keep this in fetchMarkets for now, the rule of thumb is to keep in one place everything that is related.

Reply:
yes, with `symbolDefine` i didn't mean to be separated in the same class, i just thought it could be a base method. however, will remove it now.